### PR TITLE
Stop criteria

### PR DESCRIPTION
**Pull request by nurtas-m**
Add new stop criteria. Previously number of the allowable number of epochs without improvement was fixed. Training stopped if no improvement was seen during this window. New stop criteria include possibility of increasing of that window depending on the number of epochs needed for previous improvements during training.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -121,6 +121,14 @@ And, if you want to start training from scratch:
...
 ```
 
+To reproduce the following results, set:
+```
+# for the model with size 64:
+--max_steps 80000
+# for the model with size 512:
+--max_steps 150000
+```
+
```
Comment (nshmyrev) [inline README.md]: Should work by default, not with options.

File: g2p_seq2seq/g2p.py [hunk 2]
```diff
@@ -212,11 +212,12 @@ def train(self):
 
     # This is the training loop.
     step_time, train_loss = 0.0, 0.0
-    current_step, num_iter_wo_improve = 0, 0
-    prev_train_losses, prev_valid_losses = [], []
-    num_iter_cover_train = int(sum(train_bucket_sizes) /
-                               self.params.batch_size /
-                               self.params.steps_per_checkpoint)
+    current_step, iter_inx, num_epochs_last_impr, max_num_epochs,\
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: "inx" is unclear abbreviation.

File: g2p_seq2seq/g2p.py [hunk 3]
```diff
@@ -212,11 +212,12 @@ def train(self):
...
+      num_up_trends, num_down_trends = 0, 0, 0, 2, 0, 0
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: "trend" is not a proper word here.

File: g2p_seq2seq/g2p.py [hunk 4]
```diff
@@ -212,11 +212,12 @@ def train(self):
...
+    prev_train_losses, prev_valid_losses, prev_epoch_valid_losses = [], [], []
+    num_iter_cover_train = max(1, int(sum(train_bucket_sizes) /
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: This is called "epoch", no? "cover" is not a good word in this context.

File: g2p_seq2seq/g2p.py [hunk 5]
```diff
@@ -232,45 +233,70 @@ def train(self):
...
+        # and then make a decision to continue/stop training.
+        if (iter_inx > 0
+            and iter_inx % num_iter_cover_train == 0):
+          # Calculate average validation loss during the previous epoch
+          epoch_eval_loss = self.__calc_epoch_loss(
+            prev_valid_losses[-num_iter_cover_train:])
+          if len(prev_epoch_valid_losses) > 0:
+            print('Previous min epoch eval loss: %f, current epoch eval loss: %f' %
+                  (min(prev_epoch_valid_losses), epoch_eval_loss))
+            # Check if there was improvement during last epoch
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: an improvement

File: g2p_seq2seq/g2p.py [hunk 6]
```diff
@@ -232,45 +233,70 @@ def train(self):
...
+            if (epoch_eval_loss < min(prev_epoch_valid_losses)):
+              if num_epochs_last_impr > max_num_epochs/1.5:
+                max_num_epochs = int(1.5 * num_epochs_last_impr)
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: 1.5 must be separated into something standalone and properly named, not used in multiple places in the code without name.

File: g2p_seq2seq/g2p.py [hunk 7]
```diff
@@ -316,6 +342,26 @@ def __calc_eval_loss(self):
     return eval_loss
 
 
+  def __calc_epoch_loss(self, epoch_losses):
+    """Calculate average loss during the epoch.
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Comment is wrong, this is not really an average.

File: g2p_seq2seq/g2p.py [hunk 8]
```diff
@@ -232,45 +232,62 @@ def train(self):
...
+            # max_num_epochs epochs
+            if num_epochs_last_impr > max_num_epochs:
+              break
+
+          prev_epoch_valid_losses.append(round(epoch_eval_loss, 3))
 
         prev_train_losses.append(train_loss)
         prev_valid_losses.append(eval_loss)
         step_time, train_loss = 0.0, 0.0
+        iter_idx += 1
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Use current step instead of iter_idx

File: g2p_seq2seq/g2p.py [hunk 9]
```diff
@@ -232,45 +232,62 @@ def train(self):
...
+            # Check if there was an improvement during last epoch
+            if (epoch_eval_loss < min(prev_epoch_valid_losses)):
+              if num_epochs_last_impr > max_num_epochs/window_scale:
+                max_num_epochs = int(window_scale * num_epochs_last_impr)
+              print('Improved during last epoch.')
+              prev_min_level = prev_epoch_valid_losses[-1]
+              num_epochs_last_impr = 0
+            else:
+              print('No improvement during last epoch.')
+              num_epochs_last_impr += 1
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: epochs_without_improvement

File: g2p_seq2seq/g2p.py [hunk 10]
```diff
@@ -299,23 +316,43 @@ def __calc_step_loss(self, train_buckets_scale):
...
-    eval_loss, num_iter_total = 0.0, 0.0
+    eval_loss, iter_total = 0.0, 0.0
     for bucket_id in xrange(len(self._BUCKETS)):
-      num_iter_cover_valid = int(math.ceil(len(self.valid_set[bucket_id])/
+      iter_per_valid = int(math.ceil(len(self.valid_set[bucket_id])/
                                            self.params.batch_size))
-      num_iter_total += num_iter_cover_valid
-      for batch_id in xrange(num_iter_cover_valid):
+      iter_total += iter_per_valid
+      for batch_id in xrange(iter_per_valid):
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Use xrange with batch_size step.

File: g2p_seq2seq/g2p.py [hunk 11]
```diff
@@ -299,23 +316,43 @@ def __calc_step_loss(self, train_buckets_scale):
...
     """
-    eval_loss, num_iter_total = 0.0, 0.0
+    eval_loss, iter_total = 0.0, 0.0
     for bucket_id in xrange(len(self._BUCKETS)):
-      num_iter_cover_valid = int(math.ceil(len(self.valid_set[bucket_id])/
+      iter_per_valid = int(math.ceil(len(self.valid_set[bucket_id])/
                                            self.params.batch_size))
-      num_iter_total += num_iter_cover_valid
-      for batch_id in xrange(num_iter_cover_valid):
+      iter_total += iter_per_valid
```
Comment (nshmyrev) [inline g2p_seq2seq/g2p.py]: Count iter_total in inner loop
